Финальный проект спринта №8:

KanID92 opened the pull request:
- добавить поля продолжительности задачи, даты и времени начала задачи;
- добавлен методы расчета даты и времени завершения задачи;
- добавить поле даты и времени завершения задачи для Epic;
- добавить список приоритета (по стартовому времени), добавить геттер на данный список;
- преобразовать некоторые циклы в методах с использованием Stream API;
- добавить абстрактный тестовый класс TaskManagerTest;
- дополнить тестами.

KanID92 committed:
Финальный проект спринта №8:
- добавить поля продолжительности задачи, даты и времени начала задачи;
- добавлен методы расчета даты и времени завершения задачи;
- добавить поле даты и времени завершения задачи для Epic;
- добавить список приоритета (по стартовому времени), добавить геттер на данный список;
- преобразовать некоторые циклы в методах с использованием Stream API;
- добавить абстрактный тестовый класс TaskManagerTest;
- дополнить тестами.

KanID92 committed:
Финальный проект спринта №8:
- добавить поля продолжительности задачи, даты и времени начала задачи;
- добавлен методы расчета даты и времени завершения задачи;
- добавить поле даты и времени завершения задачи для Epic;
- добавить список приоритета (по стартовому времени), добавить геттер на данный список;
- преобразовать некоторые циклы в методах с использованием Stream API;
- добавить абстрактный тестовый класс TaskManagerTest;
- дополнить тестами.

kesch9 commented on src/model/Epic.java:
```diff
@@ -1,11 +1,14 @@
 
 public class Epic extends Task {
 
+    protected LocalDateTime endTime = LocalDateTime.now();
```
Попробуйте поработать с null и в том числе и у задач, вместо того, чтобы проставлять now здесь и MAX и ZERO у задач. Выглядит не логично при создании эпика без задач, будет start максимальным значением, end текущая дата и duration ноль, не задана дата пусть будет null. Только это надо учесть везде, что может быть null

KanID92 replied:
Откорректировано.

kesch9 commented on src/model/Task.java:
```diff
@@ -8,13 +11,26 @@ public class Task {
     protected String description;
     protected int id;
     protected Progress progress;
+    protected LocalDateTime startTime = LocalDateTime.MAX; //TODO
```
Поставили TODO нужно доработать, переделайте на null

KanID92 replied:
Откорректировано.

kesch9 commented on src/model/Task.java:
```diff
@@ -55,13 +87,25 @@ public void setProgress(Progress progress) {
+        this.duration = duration;
+    }
+
+    public void setEndTime(LocalDateTime endTime) {
```
Обязательно
setEndTime не нужен, он нужен только у эпика, здесь удаляем

KanID92 replied:
Откорректировано.

kesch9 commented on src/model/Task.java:
```diff
@@ -55,13 +87,25 @@ public void setProgress(Progress progress) {
         Task task = (Task) o;
-        return id == task.id && Objects.equals(name, task.name) && Objects.equals(description, task.description) && progress == task.progress;
+        return id == task.id && Objects.equals(name, task.name) && Objects.equals(description, task.description)
+                && progress == task.progress;
```
Обязательно
Нужно добавить и остальные поля в equals и hashcode

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -206,16 +218,21 @@ public static Task fromString(String value) {
                 task = new Epic(name, description);
                 task.setId(id);
                 task.setProgress(progress);
+                task.setEndTime(endTime);
```
Обязательно
Нужно у эпика проставить не только endTime, но startTime и duration

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -157,7 +163,8 @@ static FileBackedTaskManager loadFromFile(File autoSaveFile) { // восстан
                 fileBTManager.setCounter(maxId); // восстановление счетчика id
 
-                for (SubTask subTask : fileBTManager.getSubtaskList()) { // восстановление данных об подзадачах в эпиках
+                // восстановление данных об подзадачах в эпиках
```
Обязательно
Не восстанавливается сортированный список

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -102,21 +107,32 @@ public int createEpic(Epic epic) { // Создание Epic'а.
         if (tasks.containsKey(task.getId())) {
-            tasks.put(task.getId(), task);
+            if (!isTasksIntersect(task)) {
+                tasks.put(task.getId(), task);
```
Обязательно
При обновлении задач и подзадач, нужно обновить запись в отсортированном списке, надо сначала удалить старую запись, а затем добавить новую

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -143,11 +162,13 @@ public void updateSubtask(SubTask subTask) { // Обновление подза
         deleteTasksFromHistory(tasks);
-        tasks.clear(); // удаление самих задач;
+        tasks.clear(); // удаление задач в мапе;
+        prioritySet.clear(); // удаление задач в множестве приоритета;
```
Обязательно
Нельзя чистит весь сортированный список там еще хранятся подзадачи

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -143,11 +162,13 @@ public void updateSubtask(SubTask subTask) { // Обновление подза
 
     @Override
     public void deleteAllEpics() {  //удаление всех Эпиков
+        prioritySet.clear(); // удаление задач в множестве приоритета;
```
Обязательно
Нельзя чистит весь сортированный список там еще хранятся задачи

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -157,16 +178,20 @@ public void deleteAllEpics() {  //удаление всех Эпиков
     @Override
     public void deleteAllSubtasks() {  //удаление всех Подзадач
         deleteTasksFromHistory(subTasks);
+        prioritySet.clear(); // удаление подзадач в множестве приоритета;
```
Обязательно
Нельзя чистит весь сортированный список там еще хранятся задачи

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -157,16 +178,20 @@ public void deleteAllEpics() {  //удаление всех Эпиков
         subTasks.clear();
         for (Epic epic : epics.values()) { // удаление
             epic.removeAllSubtaskFromEpic();
+            changeEpicTimeAndDuration(epic.getId());
```
Обязательно
При удалении всех подзадач, также надо пересчитать статус у эпика, он должен стать NEW

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -229,17 +258,66 @@ private void changeEpicProgress(int epicId) {
+            }
+            epic.setStartTime(epicStartTime);
+            epic.setDuration(epicDuration);
+            epic.setEndTime(epicStartTime.plus(epicDuration));
```
Обязательно
Не правильно так считать endTime, endTime это самое позднее время у подзадач. Например у нас есть три задач с duration по 10 минут и начало в 10-00, 11-00, 12-00, по вашей логике окончание будет в 10-30, а нужно в 12-10

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -229,17 +258,66 @@ private void changeEpicProgress(int epicId) {
+        } else {
+            epic.setStartTime(null);
+            epic.setDuration(null);
+            epic.setEndTime(null);
```
Вот здесь вы сбрасываете на null, поэтому при инициализации эпика тоже работаем с null

KanID92 replied:
Откорректировано.

kesch9 reviewed (commented):
Нужно еще поработать над заданием

KanID92 committed:
Финальный проект спринта №8 (корректировка по ревью от 27.03).

KanID92 commented:
Выполнил корректировку по ревью: коммит 3326444

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -157,11 +188,17 @@ static FileBackedTaskManager loadFromFile(File autoSaveFile) { // восстан
                 }
 
+                for (Epic epic : fileBTManager.getEpicList()) {
+                    fileBTManager.changeEpicProgress(epic.getId());
```
Пересчитывать статус не нужно, вы его записываете в файл и затем восстанавливаете

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -157,11 +188,17 @@ static FileBackedTaskManager loadFromFile(File autoSaveFile) { // восстан
 
+                for (Epic epic : fileBTManager.getEpicList()) {
+                    fileBTManager.changeEpicProgress(epic.getId());
+                    fileBTManager.changeEpicTimeAndDuration(epic.getId());
```
Пересчитывать время также не нужно вы его тоже пишите и восстанавливаете

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -143,11 +170,14 @@ public void updateSubtask(SubTask subTask) { // Обновление подза
     @Override
     public void deleteAllEpics() {  //удаление всех Эпиков
+        prioritySet.removeIf(task -> task.getType() == TaskType.SUBTASK
+                || task.getType() == TaskType.EPIC);
```
Эпик не храниться в сортированном списке эта проверка лишняя, удаляем только подзадачи

KanID92 replied:
Откорректировано.

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -229,17 +271,99 @@ private void changeEpicProgress(int epicId) {
+        Duration epicDuration = Duration.ZERO;
+        LocalDateTime epicEndTime = LocalDateTime.MIN;
+        ArrayList<SubTask> epicSubtasks = getEpicSubtasks(epicId);
+        if (epicSubtasks != null) {
```
Метод changeEpicTimeAndDuration внутри класса и в них идет проверка, что такой эпик есть, поэтому на null проверять не нужно и не правильно, т.к. вам нужно проверить на пустоту и сбросить на null времена. Посмотрите как у вас сделана проверка в расчете статусов
if (!epicSubtasks.isEmpty())

KanID92 replied:
Откорректировано.

kesch9 reviewed (approved):
Перед коммитом подправьте небольшие замечания

KanID92 committed:
Финальный проект спринта №8. Дополнительные корректировки:
- метод static FileBackedTaskManager loadFromFile(File autoSaveFile). Удалить пересчет времени и статуса за ненадобностью;
- класс InMemoryTaskManager. Удалить лишнюю проверка на Epic при удалении в списке приоритета в методе deleteAllEpics();
- класс InMemoryTaskManager. Метод changeEpicTimeAndDuration(int epicId). Заменить проверку на null списка сабтасков у эпика на проверку на пустоту.
- откорректировать сеттеры и геттеры времени окончания класса Epic.
